homeassistant/components/tplink/switch.py

```diff
@@ -30,57 +31,60 @@ async def async_setup_entry(
     """Set up switches."""
     data: TPLinkData = hass.data[DOMAIN][config_entry.entry_id]
     parent_coordinator = data.parent_coordinator
-    device = cast(SmartPlug, parent_coordinator.device)
-    if not device.is_plug and not device.is_strip and not device.is_dimmer:
-        return
+    device = parent_coordinator.device
+    add_device_switch = True
+    for no_switch_module in NO_SWITCH_DEVICE_MODULES:
+        if no_switch_module in device.modules:
+            add_device_switch = False
+            break
     entities: list = []
-    if device.is_strip:
+    if add_device_switch and device.children:
         # Historically we only add the children if the device is a strip
         _LOGGER.debug("Initializing strip with %s sockets", len(device.children))
         entities.extend(
             SmartPlugSwitchChild(device, parent_coordinator, child)
             for child in device.children
         )
-    elif device.is_plug:
+    elif add_device_switch:
         entities.append(SmartPlugSwitch(device, parent_coordinator))
 
-    # this will be removed on the led is implemented
-    if hasattr(device, "led"):
-        entities.append(SmartPlugLedSwitch(device, parent_coordinator))
+    if Module.Led in device.modules:
+        entities.append(
+            SmartPlugLedSwitch(device, parent_coordinator, device.modules[Module.Led])
+        )
```

Comment:
Not worth making it more complicated for the switch platform PR.

Reply:
https://github.com/home-assistant/core/pull/117898 created to remove this in the next PR.